File: _core/packages/python/openproblems/src/openproblems/project/find_project_root.py
from __future__ import annotations

def find_project_root(path: str = ".") -> str | None:
    """
    Find the root of a Viash project

    This function will recursively search for a `_viash.yaml` file
    in the parent directories of the given path.

    Args:
        path (str): Path to a file or directory

    Returns:
        str: The path to the root of the Viash project, or None if not found
    """

    import os
    
    path = os.path.abspath(path) 

    while path != "/" and not os.path.exists(os.path.join(path, "_viash.yaml")):
        path = os.path.dirname(path)

    if path == "/":
        return None

    return path

File: _core/packages/r/openproblems/R/find_project_root.R
#' Find the root of a Viash project
#'
#' This function will recursively search for a `_viash.yaml` file
#' in the parent directories of the given path.
#'
#' @param path Path to a file or directory
#' @return The path to the root of the Viash project, or NULL if not found
#'
#' @importFrom fs path_abs path_dir path file_exists path_norm
#' @export
#' @examples
#' \dontrun{
#' find_project_root("/path/to/project/subdir")
#' }
find_project_root <- function(path = ".") {
  path <- fs::path_abs(fs::path_norm(path))

  while (path != "/" && !fs::file_exists(fs::path(path, "_viash.yaml"))) {
    path <- fs::path_dir(path)
  }

  if (path == "/") {
    return(NULL)
  }

  as.character(path)
}
